src/ray/raylet/node_manager.cc

```diff
@@ -659,6 +665,33 @@ void NodeManager::HandleJobFinished(const JobID &job_id, const JobTableData &job
   worker_pool_.HandleJobFinished(job_id);
 }
 
+void NodeManager::CheckForUnexpectedWorkerDisconnects() {
+  std::vector<std::shared_ptr<ClientConnection>> all_connections;
+  std::vector<std::shared_ptr<WorkerInterface>> all_workers =
+      worker_pool_.GetAllRegisteredWorkers();
+  all_connections.reserve(all_workers.size());
+  for (const auto &worker : all_workers) {
+    all_connections.push_back(worker->Connection());
+  }
+  for (const auto &driver : worker_pool_.GetAllRegisteredDrivers()) {
+    all_workers.push_back(driver);
+    all_connections.push_back(driver->Connection());
+  }
+
+  RAY_CHECK(all_connections.size() == all_workers.size());
+
+  // Check if there are any unexpected disconnects on the worker socket connections.
+  // This will close the connection without processing remaining messages.
+  std::vector<bool> disconnects = CheckForClientDisconnects(all_connections);
+  for (size_t i = 0; i < disconnects.size(); i++) {
+    if (disconnects[i]) {
+      std::string msg = "Worker connection closed unexpectedly.";
+      RAY_LOG(DEBUG).WithField(all_workers[i]->WorkerId()) << msg;
+      DestroyWorker(all_workers[i], rpc::WorkerExitType::SYSTEM_ERROR, msg);
```

Comment:
Same as reading eof, we should call `DisconnectClient()` instead of `DestroyWorker()`

Reply:
Agree that we need to unify a bit and currently it's a bit messy. Delta between `DisconnectClient()` and `DestroyWorker()` seems to be that `DestroyWorker()` increments a counter indicating how many workers are killed from the raylet.

Reply:
I'm not too opinionated about which one we use but `DestroyWorker` seems to be a more useful superset of functionality.

This logic should really also be in the worker pool IMO not hard coded in node manager.

I'll file a ticket to clean it up holistically (and maybe do it myself at some point)